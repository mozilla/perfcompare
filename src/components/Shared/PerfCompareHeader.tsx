import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';

import { useAppSelector } from '../../hooks/app';
import { Strings } from '../../resources/Strings';
import { HeaderStyles } from '../../styles';
import ToggleDarkMode from './ToggleDarkModeButton';

interface PerfCompareHeaderProps {
  view: 'search' | 'compare-results';
}

const strings = Strings.components.header;

function PerfCompareHeader({ view }: PerfCompareHeaderProps) {
  const themeMode = useAppSelector((state) => state.theme.mode);
  const styles = HeaderStyles(themeMode, view);

  return (
    <Grid className={`header-container ${styles.container}`}>
      <ToggleDarkMode />
      <Box className='header-text'>
        <Typography
          variant='h1'
          component='div'
          align='center'
          gutterBottom
          className='perfcompare-header'
        >
          {strings.title}
        </Typography>
        {view === 'search' && (
          <>
            <Typography
              component='div'
              align='center'
              gutterBottom
              className='perfcompare-tagline'
            >
              {strings.tagline}
            </Typography>
            <Button className='learn-more-btn'>{strings.learnMore}</Button>
          </>
        )}
      </Box>
    </Grid>
  );
}

export default PerfCompareHeader;
