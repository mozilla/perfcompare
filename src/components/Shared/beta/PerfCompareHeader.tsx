import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';
import { style } from 'typestyle';

import { Strings } from '../../../resources/Strings';
import {
  Spacing,
  ButtonsLightRaw,
  ButtonsDarkRaw,
  Colors,
} from '../../../styles';
import ToggleDarkMode from './ToggleDarkModeButton';

const strings = Strings.components.header;

function PerfCompareHeader(props: PerfCompareHeaderProps) {
  const { themeMode, toggleColorMode } = props;

  const styles = {
    container: style({
      marginBottom: `${Spacing.layoutLarge + 20}px`,
      padding: 0,
      width: '100%',
      minHeight: '357px',
      backgroundColor:
        themeMode == 'light' ? Colors.Background200 : Colors.Background200Dark,
      backgroundImage: strings.bgLink,
      backgroundPosition: 'center',
      backgroundRepeat: 'no-repeat',
      backgroundPositionY: 'top',
      $nest: {
        '.header-text': {
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'center',
          maxWidth: '472px',
          margin: '0 auto',
        },
        '.perfcompare-header': {
          marginBottom: `${Spacing.Large}px`,
        },
        '.perfcompare-tagline': {
          marginBottom: `${Spacing.xxLarge}px`,
        },
        '.MuiButtonBase-root': {
          marginBottom: `${Spacing.layoutLarge + 14}px`,
          maxWidth: '104px',
          alignSelf: 'center',
          ...(themeMode == 'light'
            ? ButtonsLightRaw.Secondary
            : ButtonsDarkRaw.Secondary),
        },
        '.learn-more-btn': {
          //hidden until purpose is determined
          //visibility: 'hidden',
        },
      },
    }),
  };
  return (
    <Grid className={`header-container ${styles.container}`}>
      <ToggleDarkMode theme={themeMode} toggleColorMode={toggleColorMode} />
      <Box className='header-text'>
        <Typography
          variant='h1'
          component='div'
          align='center'
          gutterBottom
          className='perfcompare-header'
        >
          {strings.title}
        </Typography>
        <Typography
          component='div'
          align='center'
          gutterBottom
          className='perfcompare-tagline'
        >
          {strings.tagline}
        </Typography>
        <Button className='learn-more-btn'>{strings.learnMore}</Button>
      </Box>
    </Grid>
  );
}

interface PerfCompareHeaderProps {
  themeMode: 'light' | 'dark';
  toggleColorMode: () => void;
}

export default PerfCompareHeader;
