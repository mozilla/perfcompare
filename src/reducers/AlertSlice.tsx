import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import type { AlertState, AlertType } from '../types/state';

const initialState: AlertState = {
  isAlert: false,
  alert: {
    message: '',
    title: undefined,
    severity: undefined,
  },
};

const alert = createSlice({
  name: 'alert',
  initialState,
  reducers: {
    clearAlert(state) {
      state.isAlert = initialState.isAlert;
      state.alert = initialState.alert;
    },
    setAlert(state, action: PayloadAction<AlertType>) {
      state.isAlert = true;
      state.alert = action.payload;
    },
  },
});

export const { clearAlert, setAlert } = alert.actions;
export default alert.reducer;
