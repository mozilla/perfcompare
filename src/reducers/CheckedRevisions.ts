import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { CheckedRevisionsState } from '../types/state';
import type { Revision } from '../types/state';

const initialState: CheckedRevisionsState = {
  revisions: [],
};

const checkedRevisions = createSlice({
  name: 'checkedRevisions',
  initialState,
  reducers: {
    clearCheckedRevisions(state) {
      state.revisions = initialState.revisions;
    },
    setCheckedRevisions(state, action: PayloadAction<Revision[]>) {
      state.revisions = action.payload;
    },
  },
});

export const { clearCheckedRevisions, setCheckedRevisions } =
  checkedRevisions.actions;
export default checkedRevisions.reducer;
