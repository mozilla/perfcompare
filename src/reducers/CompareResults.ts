import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { fetchCompareResults } from '../thunks/compareResultsThunk';
import type { CompareResultsState, ResultsHashmap } from '../types/state';

const initialState: CompareResultsState = {
  data: {},
  loading: false,
  error: undefined,
};

const compareResults = createSlice({
  name: 'compareResults',
  initialState,
  reducers: {
    setCompareData(
      state,
      action: PayloadAction<{
        data: ResultsHashmap;
      }>,
    ) {
      state.data = action.payload.data;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchCompareResults.fulfilled, (state, action) => {
        const revisionHash = action.payload[0].new_rev;
        state.data[revisionHash] = action.payload;
        state.loading = initialState.loading;
      })
      .addCase(fetchCompareResults.pending, (state) => {
        state.loading = true;
        state.error = initialState.error;
      })
      .addCase(fetchCompareResults.rejected, (state, action) => {
        state.error = action.payload;
        state.loading = initialState.loading;
      });
  },
});

export const { setCompareData } = compareResults.actions;
export default compareResults.reducer;
