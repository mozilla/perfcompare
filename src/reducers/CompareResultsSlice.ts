import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import {
  fetchCompareResults,
  fetchFakeResults,
} from '../thunks/compareResultsThunk';
import type {
  CompareResultsState,
  CompareResultsItem,
  ResultsHashmap,
} from '../types/state';

const initialState: CompareResultsState = {
  data: {} as ResultsHashmap,
  loading: false,
  error: undefined,
};

function fetchPending(state: CompareResultsState) {
  state.loading = true;
  state.error = initialState.error;
}

function fetchFulfilled(
  state: CompareResultsState,
  action: PayloadAction<CompareResultsItem[]>,
) {
  const revisionHash = action.payload[0].new_rev;
  state.data[revisionHash] = action.payload;
  state.loading = initialState.loading;
}

function fetchRejected(
  state: CompareResultsState,
  action: PayloadAction<string | undefined>,
) {
  state.error = action.payload;
  state.loading = initialState.loading;
}

const compareResults = createSlice({
  name: 'compareResults',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchCompareResults.pending, fetchPending)
      .addCase(fetchFakeResults.pending, fetchPending)
      .addCase(fetchCompareResults.fulfilled, fetchFulfilled)
      .addCase(fetchFakeResults.fulfilled, fetchFulfilled)
      .addCase(fetchCompareResults.rejected, fetchRejected)
      .addCase(fetchFakeResults.rejected, fetchRejected);
  },
});

export default compareResults.reducer;
