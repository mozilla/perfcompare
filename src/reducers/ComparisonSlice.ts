import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit';

import { RootState } from '../common/store';
import { Strings } from '../resources/Strings';
import { CompareResultsItem, RevisionsHeader } from '../types/state';
import { truncateHash } from '../utils/helpers';

interface InitialState {
  activeComparison: string;
}

const initialState: InitialState = {
  activeComparison:
    Strings.components.comparisonRevisionDropdown.allRevisions.key,
};

type Results = {
  key: string;
  value: CompareResultsItem[];
  revisionHeader: RevisionsHeader;
};

type ResultsGroupedByKey = Record<string, ResultObject[]>;

interface ResultObject {
  header_name: string;
  new_rev: string;
}

const comparison = createSlice({
  name: 'comparison',
  initialState,
  reducers: {
    updateComparison(
      state,
      action: PayloadAction<{
        activeComparison: string;
      }>,
    ) {
      state.activeComparison = action.payload.activeComparison;
    },
  },
});

/*This function transforms results into an array of objects, where each object represents a array of objects 
grouped by their header_name as key. The keys in the resulting objects are composed of header_name and a 
truncated hash of new_rev.
for example: [
  {
    "a11yr opt e10s fission stylo webrender 69d5beb77da0": [
      {
        "base_rev": "8f5a11c1eb0b7598d1415f6efa9c360191a423f8",
        "new_rev": "69d5beb77da06f9eda78eff3c54463273d457e66",
        "framework_id": 1,...}{}{}...]}{}{}...] */
export const formatDownloadData = (
  data: Record<string, ResultObject[]>,
): Array<ResultsGroupedByKey> => {
  const groupNames = Object.keys(data);
  const transformedGroups: Array<ResultsGroupedByKey> = [];

  groupNames.forEach((groupName) => {
    const groupedResults = data[groupName].reduce(
      (grouped: ResultsGroupedByKey, result: ResultObject) => {
        if (!grouped[result.header_name]) {
          grouped[result.header_name] = [];
        }
        grouped[result.header_name].push(result);
        return grouped;
      },
      {},
    );

    const transformedGroupEntries = Object.keys(groupedResults).map(
      (header_name) => {
        const key = `${header_name} ${truncateHash(
          groupedResults[header_name][0].new_rev,
        )}`;
        const value = groupedResults[header_name];

        return {
          [key]: value,
        };
      },
    );

    transformedGroups.push(...transformedGroupEntries);
  });
  return transformedGroups;
};

export const selectStringifiedJsonResults = createSelector(
  (state: RootState) => state.comparison.activeComparison,
  (state: RootState) => state.compareResults.data,
  (activeComparison, data) => {
    if (
      activeComparison ===
      Strings.components.comparisonRevisionDropdown.allRevisions.key
    ) {
      return JSON.stringify(formatDownloadData(data), null, 2);
    } else {
      return JSON.stringify(
        formatDownloadData({
          [activeComparison]: data[activeComparison],
        }),
        null,
        2,
      );
    }
  },
);

function processResults(results: CompareResultsItem[]) {
  const processedResults: Map<string, CompareResultsItem[]> = new Map<
    string,
    CompareResultsItem[]
  >();
  results.forEach((result) => {
    const { new_rev: newRevision, header_name: header } = result;
    const rowIdentifier = header.concat(' ', newRevision);
    if (processedResults.has(rowIdentifier)) {
      (processedResults.get(rowIdentifier) as CompareResultsItem[]).push(
        result,
      );
    } else {
      processedResults.set(rowIdentifier, [result]);
    }
  });
  const restructuredResults: Results[] = Array.from(
    processedResults,
    function ([rowIdentifier, result]) {
      return {
        key: rowIdentifier,
        value: result,
        revisionHeader: {
          suite: result[0].suite,
          framework_id: result[0].framework_id,
          test: result[0].test,
          option_name: result[0].option_name,
          extra_options: result[0].extra_options,
          new_rev: result[0].new_rev,
          new_repo: result[0].new_repository_name,
        },
      };
    },
  );

  return restructuredResults;
}

const allRevisionsOption =
  Strings.components.comparisonRevisionDropdown.allRevisions.key;

export const selectProcessedResults = createSelector(
  (state: RootState) => state.compareResults.data,
  (state: RootState) => state.comparison.activeComparison,
  (data, activeComparison) => {
    if (activeComparison === allRevisionsOption) {
      const allResults = ([] as CompareResultsItem[]).concat(
        ...Object.values(data),
      );
      return processResults(allResults);
    } else {
      const results = data[activeComparison];
      return processResults(results);
    }
  },
);

export const selectNewRevisions = createSelector(
  (state: RootState) => state.selectedRevisions.newCommittedRevisions,
  (newSelectedRevisions) => {
    return newSelectedRevisions.map((item) => item.revision);
  },
);

export const { updateComparison } = comparison.actions;

export default comparison.reducer;
