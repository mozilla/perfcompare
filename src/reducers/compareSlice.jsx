import { createSlice } from '@reduxjs/toolkit'

const initialState = [
    { id: 1, revisionID: null, author: '', commitMessage: '', date: null, selected: false },
    { id: 2, revisionID: null, author: '', commitMessage: '', date: null, selected: false },
    { id: 3, revisionID: null, author: '', commitMessage: '', date: null, selected: false },
    { id: 4, revisionID: null, author: '', commitMessage: '', date: null, selected: false },
]

export const compareSlice = createSlice({
    name: 'compare',
    initialState,
    reducers: {},
})
