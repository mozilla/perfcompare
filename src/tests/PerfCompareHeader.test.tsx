import userEvent from '@testing-library/user-event';

import PerfCompareHeader from '../components/Shared/PerfCompareHeader';
import { setAlert } from '../reducers/AlertSlice';
import type { AlertType } from '../types/state';
import { render, screen, store } from './utils/test-utils';

const message = 'Maximum 4 Revisions';
const severity: AlertType['severity'] = 'error';
let title;

describe('PerfCompareHeader', () => {
  it('renders correctly when there is no alert', () => {
    render(<PerfCompareHeader />);

    expect(document.body).toMatchSnapshot();
  });

  it('should hide alert when close button is clicked', async () => {
    // set delay to null to prevent test time-out due to useFakeTimers
    const user = userEvent.setup({ delay: null });
    title = undefined;

    store.dispatch(setAlert({ message, severity, title }));
    render(<PerfCompareHeader />);

    expect(screen.getByText('Maximum 4 Revisions')).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();

    await user.click(screen.getByRole('button', { name: 'Close' }));
    expect(screen.queryByText('Maximum 4 Revisions')).not.toBeInTheDocument();
  });

  it('should should show alert title', async () => {
    title = 'This is a title';

    store.dispatch(setAlert({ message, severity, title }));
    render(<PerfCompareHeader />);

    expect(screen.getByText('This is a title')).toBeInTheDocument();

    expect(document.body).toMatchSnapshot();
  });
});
