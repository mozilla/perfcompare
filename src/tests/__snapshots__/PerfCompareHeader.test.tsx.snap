// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PerfCompareHeader renders correctly when there is no alert 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-5q2xqw-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
        />
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
        />
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
        />
      </div>
      <div
        class="MuiTypography-root MuiTypography-h1 MuiTypography-alignCenter MuiTypography-gutterBottom perfcompare-header css-1bn4tuo-MuiTypography-root"
      >
        PerfCompare
      </div>
    </div>
  </div>
</body>
`;

exports[`PerfCompareHeader should hide alert when close button is clicked 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-5q2xqw-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
        />
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-xilmwd-MuiPaper-root-MuiAlert-root"
            role="alert"
          >
            <div
              class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="ErrorOutlineIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
            </div>
            <div
              class="MuiAlert-message css-acap47-MuiAlert-message"
            >
              Maximum 4 Revisions
            </div>
            <div
              class="MuiAlert-action css-ki1hdl-MuiAlert-action"
            >
              <button
                aria-label="Close"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall css-1e0d89p-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                title="Close"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                  data-testid="CloseIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
        />
      </div>
      <div
        class="MuiTypography-root MuiTypography-h1 MuiTypography-alignCenter MuiTypography-gutterBottom perfcompare-header css-1bn4tuo-MuiTypography-root"
      >
        PerfCompare
      </div>
    </div>
  </div>
</body>
`;

exports[`PerfCompareHeader should should show alert title 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-5q2xqw-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
        />
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-xilmwd-MuiPaper-root-MuiAlert-root"
            role="alert"
          >
            <div
              class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="ErrorOutlineIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
            </div>
            <div
              class="MuiAlert-message css-acap47-MuiAlert-message"
            >
              <div
                class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiAlertTitle-root css-yslrua-MuiTypography-root-MuiAlertTitle-root"
              >
                This is a title
              </div>
              Maximum 4 Revisions
            </div>
            <div
              class="MuiAlert-action css-ki1hdl-MuiAlert-action"
            >
              <button
                aria-label="Close"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall css-1e0d89p-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                title="Close"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                  data-testid="CloseIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
        />
      </div>
      <div
        class="MuiTypography-root MuiTypography-h1 MuiTypography-alignCenter MuiTypography-gutterBottom perfcompare-header css-1bn4tuo-MuiTypography-root"
      >
        PerfCompare
      </div>
    </div>
  </div>
</body>
`;
